compositor_render/src/scene/transition/cubic_bezier.rs

```diff
@@ -0,0 +1,146 @@
+}
+
+fn cubic_bezier(t: f64, p1: f64, p2: f64) -> f64 {
+    let a = 1.0 / 3.0 + (p1 - p2);
+    let b = p2 - 2.0 * p1;
+    let c = p1;
+
+    3.0 * ((a * t + b) * t + c) * t
+}
+
+// Based on https://pomax.github.io/bezierinfo/#yforx
+// https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/ui/ui-graphics/src/commonMain/kotlin/androidx/compose/ui/graphics/Bezier.kt;l=274;drc=9b09524f2b2cbfc397269fbbe1267dd46b14077a
+fn find_first_cubic_root(p0: f64, p1: f64, p2: f64, p3: f64) -> f64 {
+    // Coefficients calculated from control points
+    let a = 3.0 * (p0 - 2.0 * p1 + p2);
+    let b = 3.0 * (p1 - p0);
+    let c = p0;
+    let d = -p0 + 3.0 * (p1 - p2) + p3;
+
+    // If `d` is 0 then the function is not cubic
+    if d.is_close_to(0.0) {
+        // If `a` is 0 then the function is not quadratic
+        if a.is_close_to(0.0) {
+            // If `b` is 0 then the function is not linear
+            if b.is_close_to(0.0) {
+                return f64::NAN;
+            }
+
+            return (-c / b).clamp_valid_root_in_unit_range();
+        }
+        let q = f64::sqrt(b * b - 4.0 * a * c);
+        let a2 = 2.0 * a;
+        let root = ((q - b) / a2).clamp_valid_root_in_unit_range();
+        if !root.is_nan() {
+            return root;
+        }
+
+        return ((-b - q) / a2).clamp_valid_root_in_unit_range();
+    }
+
+    let a = a / d;
+    let b = b / d;
+    let c = c / d;
+
+    let o3 = (3.0 * b - a.powi(2)) / 9.0;
+    let q2 = (2.0 * a.powi(3) - 9.0 * a * b + 27.0 * c) / 54.0;
+    let a3 = a / 3.0;
+    let discriminant = q2.powi(2) + o3.powi(3);
+
+    if discriminant < 0.0 {
+        let mp33 = -f64::powi(o3, 3);
+        let r = mp33.sqrt();
+        let cos_phi = (-q2 / r).clamp(-1.0, 1.0);
+        let phi = cos_phi.acos();
+        let t1 = 2.0 * r.cbrt();
+
+        let root = (t1 * f64::cos(phi / 3.0) - a3).clamp_valid_root_in_unit_range();
+        if !root.is_nan() {
+            return root;
+        }
+
+        let root = (t1 * f64::cos((phi + 2.0 * PI) / 3.0) - a3).clamp_valid_root_in_unit_range();
+        if !root.is_nan() {
+            return root;
+        }
+
+        return (t1 * f64::cos((phi + 4.0 * PI) / 3.0) - a3).clamp_valid_root_in_unit_range();
+    }
+
+    if discriminant == 0.0 {
+        let u1 = -f64::cbrt(q2);
+        let root = (2.0 * u1 - a3).clamp_valid_root_in_unit_range();
+        if !root.is_nan() {
+            return root;
+        }
+
+        return (-u1 - a3).clamp_valid_root_in_unit_range();
+    }
+
+    let sd = discriminant.sqrt();
+    let u1 = (-q2 + sd).cbrt();
+    let v1 = (q2 + sd).cbrt();
+
+    (u1 - v1 - a3).clamp_valid_root_in_unit_range()
+}
+
+trait F64Ext {
+    fn is_close_to(&self, value: Self) -> bool;
+
+    fn clamp_valid_root_in_unit_range(self) -> Self;
+}
+
+impl F64Ext for f64 {
+    fn is_close_to(&self, value: Self) -> bool {
+        (*self - value).abs() < 2.0 * f64::EPSILON
```

Comment:
are you sure these comparisons involving `f64::EPSILON` are correct? `f64::EPSILON` is the difference between 1.0 and the next representable number. I believe this means that there are no representable numbers (other than `2.0`) in the range `(2.0 - f64::EPSILON, 2.0 + f64::EPSILON)`, since the representable floats get more and more spread apart the further away from `0.0` you are. For a big part of all numbers, this comparison is equivalent to a simple `==` comparison. Am I wrong here? Is this the effect you wanted to achieve?

Reply:
In the cases I tested, `2 * Epsilon` was enough. But I agree that it could be larger. I increased the allowed error to `1e-7`